consensus/core/src/block_manager.rs

```diff
@@ -61,22 +72,65 @@ impl BlockManager {
-                unsuspended_blocks.push(block);
+            if let Some(block) = self.try_accept_one_block(block) {
+                // Try to unsuspend any children blocks.
+                let unsuspended_blocks = self.try_unsuspend_children_blocks(&block);
+
+                // Try to verify the block with ancestor blocks.
+                let mut blocks_to_accept: BTreeMap<BlockRef, VerifiedBlock> = BTreeMap::new();
+                let mut blocks_to_reject: BTreeMap<BlockRef, VerifiedBlock> = BTreeMap::new();
+                {
+                    'block: for b in iter::once(block).chain(unsuspended_blocks) {
+                        let ancestors = self.dag_state.read().get_blocks(b.ancestors());
+                        assert_eq!(b.ancestors().len(), ancestors.len());
+                        let mut ancestor_blocks = vec![];
+                        'ancestor: for (included, found) in
+                            b.ancestors().iter().zip(ancestors.into_iter())
+                        {
+                            // When block is found, this should be guaranteed by the DagState.
+                            if let Some(found_block) = found {
+                                assert_eq!(included, &found_block.reference());
+                                ancestor_blocks.push(found_block);
+                                continue 'ancestor;
+                            }
+                            // blocks_to_accept have not been added to DagState yet, but they
+                            // can appear in ancestors.
+                            if blocks_to_accept.contains_key(included) {
+                                ancestor_blocks.push(blocks_to_accept[included].clone());
+                                continue 'ancestor;
+                            }
+                            // If an ancestor is already rejected, reject this block as well.
+                            if blocks_to_reject.contains_key(included) {
+                                blocks_to_reject.insert(b.reference(), b);
+                                continue 'block;
+                            }
+                            panic!("Unsuspended block {:?} has missing ancestors! Block not found in DagState: {:?}", b, included);
+                        }
+                        if let Err(e) = self.block_verifier.check_ancestors(&b, &ancestor_blocks) {
+                            warn!("Block {:?} failed to verify ancestors: {}", b, e);
+                            blocks_to_reject.insert(b.reference(), b);
+                        } else {
+                            blocks_to_accept.insert(b.reference(), b);
+                        }
+                    }
+                }
 
-                // Accept the state in DAG here so the next block to be processed can find in DAG/store any accepted blocks.
+                // TODO: report blocks_to_reject to peers.
```

Comment:
Yes, and these are not just suspected but verifiable byzantine behaviors, because of block signature. Only the block without a rejected ancestor can be reported though.